File: Lab6/Grammar.java
import domain.Pair;

import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class Grammar {

    private List<String> N;
    private List<String> E;
    private List<Pair<String, String>> P;

    public List<String> lineSplitter(String line) {
        return Arrays.stream(line.strip().split(" ")).skip(2L).collect(Collectors.toList());
    }

    public Grammar() {
        N = new ArrayList<>();
        E = new ArrayList<>();
        P = new ArrayList<>();
    }

    public List<String> getN() {
        return N;
    }

    public List<String> getE() {
        return E;
    }

    public List<Pair<String, String>> getP() {
        return P;
    }

    public List<Pair<String, String>> filterP(String nonTerminal){
        return P.stream().filter(pair-> pair.getKey().equals(nonTerminal)).collect(Collectors.toList());
    }

    public void readGrammar(String fileName) throws Exception {
        try (FileReader fileReader = new FileReader(fileName)) {

            BufferedReader bufferedReader =
                    new BufferedReader(fileReader);
            String line = bufferedReader.readLine();
            N = lineSplitter(line);
            line = bufferedReader.readLine();
            E = lineSplitter(line);
            bufferedReader.readLine();
            line = bufferedReader.readLine();
            while (line != null) {
                List<String> tokens = Arrays.stream(line.strip()
                        .replace("->", " ")
                        .split(" "))
                        .collect(Collectors.toList());
                if (N.contains(tokens.get(0))) {
                    Pair<String,String>  pair= new Pair( tokens.get(0), tokens.get(1));
                    P.add(pair);
                    line = bufferedReader.readLine();
                } else {
                    throw new Exception("Nonterminal doesn't exist");
                }
            }
        }
    }
}

File: Lab6 redone/Grammar.py
class Pair:
    def __init__(self, first, second):
        self._first = first
        self._second = second

    @property
    def first(self):
        return self._first

    @property
    def second(self):
        return self._second

    def __repr__(self):
        return "<" + str(self._first) + " | " + str(self._second) + ">"

    def __eq__(self, other):
        return self._first == other.first and self._second == other.second

class Grammar:
    def __init__(self):
        self._N = []
        self._E = []
        self._P = []

    @staticmethod
    def line_splitter(line: str) -> list:
        return line.strip().split(" ")[2:]

    @property
    def N(self) -> list:
        return self._N

    @property
    def E(self) -> list:
        return self._E

    @property
    def P(self) -> list:
        return self._P

    def filter_P(self, nonTerminal: str) -> list:
        return list(filter(lambda el: el.first == nonTerminal, self._P))

    def read_grammar(self, filename: str) -> None:
        with open(filename, 'r') as f:
            self._N = self.line_splitter(f.readline())
            self._E = self.line_splitter(f.readline())
            f.readline()
            line = f.readline()
            while line:
                tokens = line.strip().split("->")
                if tokens[0] in self._N:
                    pair = Pair(tokens[0], tokens[1])
                    self._P.append(pair)
                    line = f.readline()
                else:
                    raise Exception("NonTerminal doesn't exist!")
